File: Yeti/Controllers/Sidebar/FolderCell.m
//
//  FolderCell.m
//  Elytra
//
//  Created by Nikhil Nigade on 08/08/20.
//  Copyright © 2020 Dezine Zync Studios. All rights reserved.
//

#import "FolderCell.h"

@interface FolderCell ()

@property (nonatomic, copy, readwrite) NSNumber *folderID;

@end

@implementation FolderCell

- (void)configure:(Folder *)item indexPath:(NSIndexPath *)indexPath {
    
    self.folder = item;
    
    if (self.folder == nil) {
        return;
    }
    
    UIListContentConfiguration *content = [UIListContentConfiguration sidebarHeaderConfiguration];
    
    content.textProperties.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    content.text = item.title;
    
    if (SharedPrefs.showUnreadCounts == YES) {
        
        item.unreadCountObservor = self;
        
        if (item.unreadCount.unsignedIntegerValue > 0) {
            
            content.secondaryText = item.unreadCount.stringValue;
            
        }
        
    }
    
    if (self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomMac) {
        content.textProperties.color = UIColor.secondaryLabelColor;
    }
    else {
        content.textProperties.color = UIColor.labelColor;
    }
    
    content.secondaryTextProperties.color = UIColor.secondaryLabelColor;
    
    content.prefersSideBySideTextAndSecondaryText = YES;
    
    NSDiffableDataSourceSectionSnapshot *snapshot = [self.DS snapshotForSection:@(NSUIntegerMax - 200)];
    
    NSString *imageName = [snapshot isExpanded:item] ? @"folder" : @"folder.fill";
    
    content.image = [UIImage systemImageNamed:imageName];
    
    content.imageProperties.maximumSize = CGSizeMake(24.f, 24.f);
    
    self.contentConfiguration = content;
    
    UICellAccessoryOutlineDisclosure *disclosure = [UICellAccessoryOutlineDisclosure new];
    
    disclosure.style = UICellAccessoryOutlineDisclosureStyleCell;
    
    self.accessories = @[disclosure];
    
}

- (void)updateConfigurationUsingState:(UICellConfigurationState *)state {
    
    UIListContentConfiguration *updatedContent = (id)[self contentConfiguration];
    
    UIBackgroundConfiguration *backgroundConfig = self.backgroundConfiguration;
    
    if (self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPhone) {
        
        backgroundConfig.backgroundColor = UIColor.systemBackgroundColor;
        
        self.backgroundConfiguration = backgroundConfig;
    }
    
    if (state.isExpanded == YES) {
        
        updatedContent.image = [UIImage systemImageNamed:@"folder"];
    
    }
    else {
        
        updatedContent.image = [UIImage systemImageNamed:@"folder.fill"];
        
    }
    
//#if !TARGET_OS_MACCATALYST
    if (state.isSelected) {
        backgroundConfig.backgroundColor = UIColor.tertiarySystemFillColor;
    }
    else {
        backgroundConfig.backgroundColor = UIColor.clearColor;
    }
//#endif
    
    self.backgroundConfiguration = backgroundConfig;
    
    self.contentConfiguration = updatedContent;
    
}

- (void)unreadCountChangedFor:(Folder *)folder to:(NSNumber *)count {
    
    /*
     * in iOS 14 - Beta 4, when a cell is expanded,
     * the primary cell is hidden and replaced with
     * with a visible cell at the same index path.
     * Because of this, we cannot reference *self* here.
     */
       
    FolderCell *cell = (id)[(UICollectionView *)[self.DS valueForKey:@"collectionView"] cellForItemAtIndexPath:[self.DS indexPathForItemIdentifier:folder]];

    if (cell == nil) {
       return;
    }
    
    UIListContentConfiguration *content = (id)[cell contentConfiguration];
        
    if (count.unsignedIntegerValue > 0) {
        content.secondaryText = count.stringValue;
    }
    else {
        content.secondaryText = nil;
    }
    
    cell.contentConfiguration = content;
    
}

@end

File: Yeti/Controllers/Sidebar/FolderCell.swift
//
//  FolderCell.swift
//  Elytra
//
//  Created by Nikhil Nigade on 14/03/21.
//  Copyright © 2021 Dezine Zync Studios. All rights reserved.
//

import UIKit
import Models
import Combine

class FolderCell: UICollectionViewListCell {
    
    weak var folder: Folder?
    var item: SidebarItem?
    var indexPath: IndexPath?
    weak var DS: UICollectionViewDiffableDataSource<Int, SidebarItem>?
    
    var cancellables: [AnyCancellable] = []
    
    func configure(_ item: SidebarItem, indexPath: IndexPath) {
        
        guard case .folder(let folder) = item else {
            return
        }
        
        self.folder = folder
        self.item = item
        self.indexPath = indexPath
        
        var content = UIListContentConfiguration.sidebarHeader()
        
        content.textProperties.font = UIFont.preferredFont(forTextStyle: .body)
        
        folder.publisher(for: \.title)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] (title) in
            
            guard let sself = self else {
                return
            }
            
            CoalescingQueue.standard.add(sself, #selector(sself.updateTitle))
            
        }
        .store(in: &cancellables)
        
        if SharedPrefs.showUnreadCounts == true {
            
            folder.$unread
                .receive(on: DispatchQueue.main)
                .sink { [weak self] (unread) in
                    
                    guard let sself = self else {
                        return
                    }
                
                    CoalescingQueue.standard.add(sself, #selector(sself.updateUnreadCount))
                
            }
            .store(in: &cancellables)
            
        }
        
        content.textProperties.color = .label
        
        content.secondaryTextProperties.color = .secondaryLabel
        
        content.prefersSideBySideTextAndSecondaryText = true
        
        let expanded = DS?.snapshot(for: SidebarSection.folders.rawValue).isExpanded(item)
        
        let imageName = expanded == true ? "folder" : "folder.fill"
        
        content.image = UIImage(systemName: imageName)
        
        content.imageProperties.maximumSize = CGSize(width: 24, height: 24)
        
        contentConfiguration = content
        
        if accessories.count == 0 {
            
            let options = UICellAccessory.OutlineDisclosureOptions(style: .cell)
            let disclosure = UICellAccessory.outlineDisclosure(options: options)
            
            accessories = [disclosure]
            
        }
        
    }
    
    override func prepareForReuse() {
        
        if cancellables.count > 0 {
            
            for c in cancellables {
                c.cancel()
            }
            
            cancellables.removeAll()
            
        }
        
        super.prepareForReuse()
        
    }
    
    override func updateConfiguration(using state: UICellConfigurationState) {
        
        guard var updatedContent = self.contentConfiguration as? UIListContentConfiguration,
              var background = self.backgroundConfiguration else {
            return
        }
        
        if traitCollection.userInterfaceIdiom == .phone {
            
            background.backgroundColor = .systemBackground
            
        }
        
        if state.isExpanded == true {
            
            updatedContent.image = UIImage(systemName: "folder")
            
        }
        else {
            
            updatedContent.image = UIImage(systemName: "folder.fill")
            
        }
        
        if state.isSelected == true {
            
            background.backgroundColor = .tertiarySystemFill
            
        }
        else {
            background.backgroundColor = .clear
        }
        
        backgroundConfiguration = background
        contentConfiguration = updatedContent
        
    }
    
    @objc func updateUnreadCount () {
        
        guard var content = contentConfiguration as? UIListContentConfiguration else {
            return
        }
        
        let unread = self.folder?.unread ?? 0
        
        content.secondaryText = unread > 0 ? "\(unread)" : ""
        
        contentConfiguration = content
        
    }
    
    @objc func updateTitle() {
        
        guard var content = contentConfiguration as? UIListContentConfiguration else {
            return
        }
        
        content.text = folder!.title
        
        runOnMainQueueWithoutDeadlocking { [weak self] in
            self?.contentConfiguration = content
        }
        
    }
    
}
